File: src/arduinobot_remote/src/task_server.cpp
#include <ros/ros.h>
#include <actionlib/server/simple_action_server.h>
#include <arduinobot_remote/ArduinobotTaskAction.h>
#include <moveit/move_group_interface/move_group_interface.h>
#include <vector>

class TaskServer
{
private:
    /* data */
    ros::NodeHandle _nodeHandle;
    actionlib::SimpleActionServer<arduinobot_remote::ArduinobotTaskAction> _actionServer;
    arduinobot_remote::ArduinobotTaskResult _result;
    std::vector<double> _arm_goal;
    std::vector<double> _gripper_goal;
    std::string _actionName;
    moveit::planning_interface::MoveGroupInterface _armMoveGroup;
    moveit::planning_interface::MoveGroupInterface _gripperMoveGroup;

public:
    TaskServer(std::string name);

    void executeCb(const arduinobot_remote::ArduinobotTaskGoalConstPtr &goal);
};

TaskServer::TaskServer(std::string name) : _actionServer(_nodeHandle, name, boost::bind(&TaskServer::executeCb, this, _1), false),
                                           _actionName(name),
                                           _armMoveGroup("arduinobot_arm"),
                                           _gripperMoveGroup("arduinobot_hand")
{
    _actionServer.start();
}

void TaskServer::executeCb(const arduinobot_remote::ArduinobotTaskGoalConstPtr &goal)
{
    bool success = true;

    if (goal->task_number == 0)
    {
        _arm_goal = {0.0, 0.0, 0.0};
        _gripper_goal = {-0.7, 0.7};
    }
    else if (goal->task_number == 1)
    {
        _arm_goal = {-1.14, -0.6, -0.07};
        _gripper_goal = {0.0, 0.0};
    }
    else if (goal->task_number == 2)
    {
        _arm_goal = {-1.57, 0.0, -1.0};
        _gripper_goal = {0.0, 0.0};
    }
    else
    {
        ROS_ERROR("Invalid goal");
        return;
    }

    _armMoveGroup.setJointValueTarget(_arm_goal);
    _gripperMoveGroup.setJointValueTarget(_gripper_goal);

    _armMoveGroup.move();
    _gripperMoveGroup.move();

    _armMoveGroup.stop();
    _gripperMoveGroup.stop();

    if (_actionServer.isPreemptRequested() || !ros::ok())
    {
        ROS_INFO("%s: Preempted", _actionName.c_str());
        _actionServer.setPreempted();
        success = false;
    }

    if (success)
    {
        _result.success = true;
        ROS_INFO("%s Succeeded", _actionName.c_str());
        _actionServer.setSucceeded(_result);
    }
}

int main(int argc, char **argv)
{
    ros::init(argc, argv, "task_server");
    TaskServer server("task_server");
    ros::spin();
}

File: src/arduinobot_remote/scripts/task_server.py
#! usr/bin/env python3
import sys
import rospy
import actionlib
import moveit_commander
from arduinobot_remote.msg import ArduinobotTaskAction, ArduinobotTaskResult


class TaskServer(object):
    result = ArduinobotTaskResult()
    arm_goal = []
    gripper_goal = []

    def __init__(self, name: str) -> None:
        self.action_name = name
        moveit_commander.roscpp_initialize(sys.argv)
        self.arm_move_group = moveit_commander.MoveGroupCommander(
            "arduinobot_arm")
        self.hand_move_group = moveit_commander.MoveGroupCommander(
            "arduinobot_hand")
        self.actionServer = actionlib.SimpleActionServer(
            self.action_name, ArduinobotTaskAction, execute_cb=self.executeCb, auto_start=False)
        self.actionServer.start()

    def executeCb(self, goal):
        success = True
        if goal.task_number == 0:
            self.arm_goal = [0.0, 0.0, 0.0]
            self.gripper_goal = [-0.7, 0.7]
        elif goal.task_number == 1:
            self.arm_goal = [-1.14, -0.6, -0.07]
            self.gripper_goal = [0., 0.]
        elif goal.task_number == 2:
            self.arm_goal = [-1.57, 0., -1.0]
            self.gripper_goal = [0., 0.]
        else:
            rospy.loginfo("Invalid goal")
            return

        self.arm_move_group.go(self.arm_goal, wait=True)
        self.hand_move_group.go(self.gripper_goal, wait=True)
        self.arm_move_group.stop()
        self.hand_move_group.stop()

        if self.actionServer.is_preempt_requested():
            rospy.loginfo("%s is preempted" % self.action_name)
            self.actionServer.set_preempted()
            success = False

        if success:
            self.result.success = True
            rospy.loginfo("%s Succeeded" % self.action_name)
            self.actionServer.set_succeeded(self.result)


if __name__ == "__main__":
    rospy.init_node("task_server")
    ts = TaskServer(rospy.get_name())
    rospy.spin()
